src/qibo/backends/numpy.py

```diff
@@ -737,22 +750,22 @@ def calculate_matrix_exp(self, a, matrix, eigenvectors=None, eigenvalues=None):
             else:
                 from scipy.linalg import expm
             return expm(-1j * a * matrix)
-        else:
-            expd = self.np.diag(self.np.exp(-1j * a * eigenvalues))
-            ud = self.np.transpose(self.np.conj(eigenvectors))
-            return self.np.matmul(eigenvectors, self.np.matmul(expd, ud))
+        expd = self.np.diag(self.np.exp(-1j * a * eigenvalues))
+        ud = self.np.transpose(self.np.conj(eigenvectors))
+        return self.np.matmul(eigenvectors, self.np.matmul(expd, ud))
 
     def calculate_expectation_state(self, hamiltonian, state, normalize):
+        state = self.cast(state)
         statec = self.np.conj(state)
-        hstate = hamiltonian @ state
+        hstate = self.cast(hamiltonian @ state)
```

Comment:
Here as well: why aren't `hamiltonian` and `state` already in the correct backend?